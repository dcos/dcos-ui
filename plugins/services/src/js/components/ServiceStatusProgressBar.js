import PropTypes from "prop-types";
import React from "react";
import { Tooltip } from "reactjs-components";
/* eslint-disable no-unused-vars */
import { Trans, Plural } from "@lingui/macro";
/* eslint-enable no-unused-vars */

import ProgressBar from "#SRC/js/components/ProgressBar";
import Pod from "../structs/Pod";
import Service from "../structs/Service";
import ServiceTree from "../structs/ServiceTree";
import ServiceStatus from "../constants/ServiceStatus";

class ServiceStatusProgressBar extends React.Component {
  getTooltipContent() {
    const { service } = this.props;
    const runningInstances = service.getRunningInstancesCount();
    const instancesTotal = service.getInstancesCount();

    return (
      <div className="tooltip-line-item">
        <Plural
          render="span"
          value={runningInstances}
          one={`# instance running out of ${instancesTotal}`}
          other={`# instances running out of ${instancesTotal}`}
        />
      </div>
    );
  }

  render() {
    const { service } = this.props;
    const instancesCount = service.getInstancesCount();
    const serviceStatus = service.getServiceStatus();
    const runningInstances = service.getRunningInstancesCount();

    if (
      serviceStatus === ServiceStatus.RUNNING ||
      serviceStatus === ServiceStatus.STOPPED ||
      serviceStatus === ServiceStatus.NA
    ) {
      return null;
    }

    return (
      <Tooltip interactive={true} content={this.getTooltipContent()}>
        <ProgressBar
          className="status-bar--large staged"
          data={[
            {
              className: "success",
              value: runningInstances
            }
          ]}
          total={instancesCount}
        />
      </Tooltip>
    );
  }
}

ServiceStatusProgressBar.propTypes = {
  service: PropTypes.oneOfType([
    PropTypes.instanceOf(Service),
    PropTypes.instanceOf(ServiceTree),
    PropTypes.instanceOf(Pod)
  ]).isRequired
};

module.exports = ServiceStatusProgressBar;
