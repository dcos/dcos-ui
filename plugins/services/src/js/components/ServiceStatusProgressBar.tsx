import * as PropTypes from "prop-types";
import * as React from "react";
import { Tooltip } from "reactjs-components";

import ProgressBar from "#SRC/js/components/ProgressBar";
import Pod from "../structs/Pod";
import Service from "../structs/Service";
import ServiceTree from "../structs/ServiceTree";
import ServiceStatus from "../constants/ServiceStatus";

const { Plural } = require("@lingui/macro");

interface ServiceStatusProgressBarProps {
  service: Service | ServiceTree | Pod;
}

class ServiceStatusProgressBar extends React.Component<
  ServiceStatusProgressBarProps
> {
  static propTypes = {
    service: PropTypes.oneOfType([
      PropTypes.instanceOf(Service),
      PropTypes.instanceOf(ServiceTree),
      PropTypes.instanceOf(Pod)
    ]).isRequired
  };

  getTooltipContent() {
    const { service } = this.props;
    const runningInstances = service.getRunningInstancesCount();
    const instancesTotal = service.getInstancesCount();

    return (
      <div className="tooltip-line-item">
        <Plural
          render="span"
          value={runningInstances}
          one={`# instance running out of ${instancesTotal}`}
          other={`# instances running out of ${instancesTotal}`}
        />
      </div>
    );
  }

  render() {
    const { service } = this.props;
    const instancesCount = service.getInstancesCount();
    const serviceStatus = service.getServiceStatus();
    const runningInstances = service.getRunningInstancesCount();

    if (
      serviceStatus === ServiceStatus.RUNNING ||
      serviceStatus === ServiceStatus.STOPPED ||
      serviceStatus === ServiceStatus.NA
    ) {
      return null;
    }

    return (
      <Tooltip interactive={true} content={this.getTooltipContent()}>
        <ProgressBar
          className="status-bar--large staged"
          data={[
            {
              className: "success",
              value: runningInstances
            }
          ]}
          total={instancesCount}
        />
      </Tooltip>
    );
  }
}

module.exports = ServiceStatusProgressBar;
export default ServiceStatusProgressBar;
