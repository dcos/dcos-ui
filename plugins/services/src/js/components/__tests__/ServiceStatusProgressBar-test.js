import React from "react";
import { shallow } from "enzyme";

const Tooltip = require("reactjs-components").Tooltip;

const ProgressBar = require("#SRC/js/components/ProgressBar");
const ServiceStatusProgressBar = require("../ServiceStatusProgressBar");
const Application = require("../../structs/Application");
const Pod = require("../../structs/Pod");

const service = new Application({
  instances: 1,
  tasksRunning: 1,
  tasksHealthy: 0,
  tasksUnhealthy: 0,
  tasksUnknown: 0,
  tasksStaged: 0
});

let thisInstance;

describe("ServiceStatusProgressBar", function() {
  beforeEach(function() {
    thisInstance = shallow(<ServiceStatusProgressBar service={service} />);
  });

  describe("ProgressBar", function() {
    it("display ProgressBar", function() {
      thisInstance = shallow(
        <ServiceStatusProgressBar
          service={
            new Pod({
              env: {
                SDK_UNINSTALL: true
              },
              spec: {
                scaling: {
                  instances: 10,
                  kind: "fixed"
                }
              },
              instances: [
                {
                  status: "stable"
                },
                {
                  status: "stable"
                },
                {
                  status: "stable"
                },
                {
                  status: "stable"
                }
              ]
            })
          }
        />
      );

      const progressBar = thisInstance.find(ProgressBar);
      expect(progressBar).toBeTruthy();
      expect(progressBar.prop("total")).toBe(10);
      expect(progressBar.prop("data")[0].value).toBe(4);
    });
  });

  describe("Tooltip", function() {
    it("display Tooltip", function() {
      const app = new Application({
        queue: {
          overdue: true
        }
      });

      thisInstance = shallow(<ServiceStatusProgressBar service={app} />);

      expect(expect(thisInstance.find(Tooltip).exists())).toBeTruthy();
    });

    it("get #getTooltipContent", function() {
      const childrenContent = shallow(
        thisInstance.instance().getTooltipContent()
      )
        .find("WithI18n")
        .props().values;

      expect(childrenContent).toMatchObject({
        instancesTotal: 1,
        runningInstances: 0
      });
    });
  });
});
