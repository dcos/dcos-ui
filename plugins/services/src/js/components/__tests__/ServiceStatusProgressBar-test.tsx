import * as React from "react";
import renderer from "react-test-renderer";

import { ServiceProgressBar } from "../ServiceStatusProgressBar";

describe("ServiceProgressBar", () => {
  for (const [runningInstances, instancesCount] of [
    [0, 1],
    [0, 10],
    [3, 4],
    [10, 10],
  ]) {
    it(`displays a (${runningInstances} / ${instancesCount}) ProgressBar`, () => {
      expect(
        renderer
          .create(
            <ServiceProgressBar
              instancesCount={instancesCount}
              runningInstances={runningInstances}
            />
          )
          .toJSON()
      ).toMatchSnapshot();
    });
  }
});
