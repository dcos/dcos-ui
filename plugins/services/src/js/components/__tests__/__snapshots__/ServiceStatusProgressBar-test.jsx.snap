// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceProgressBar displays a (0 / 1) ProgressBar 1`] = `
<div
  className="tooltip-wrapper text-align-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="status-bar flex-box status-bar--large staged"
  />
</div>
`;

exports[`ServiceProgressBar displays a (0 / 10) ProgressBar 1`] = `
<div
  className="tooltip-wrapper text-align-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="status-bar flex-box status-bar--large staged"
  />
</div>
`;

exports[`ServiceProgressBar displays a (3 / 4) ProgressBar 1`] = `
<div
  className="tooltip-wrapper text-align-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="status-bar flex-box status-bar--large staged"
  >
    <span
      className="bar success"
      style={
        Object {
          "width": "75%",
        }
      }
    />
  </div>
</div>
`;

exports[`ServiceProgressBar displays a (10 / 10) ProgressBar 1`] = `
<div
  className="tooltip-wrapper text-align-center"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="status-bar flex-box status-bar--large staged"
  >
    <span
      className="bar success"
      style={
        Object {
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;
